### ISSUE TITLE
Incorrect operand order between SSE and AVX

### ISSUE THREAD

**Issue by YValeri** [DynamoRIO/dynamorio#3775]
There is currently a problem regarding the way DynamoRIO manages AVX and SSE floating point instructions, explained in this post : (https://groups.google.com/forum/#!topic/DynamoRIO-Users/EwjJLo-fBdo). For AVX, the Intel standard (https://software.intel.com/sites/default/files/managed/39/c5/325462-sdm-vol-1-2abcd-3abcd.pdf) specifies that for the instruction are as followed :
"instr destination, source0, source1", so all the operands are explicitely given. However, for SSE, the standard differs a bit : "instr destination(=source0), source1", so the destination and first source are the same, the instructions are done in-place.
Regarding DynamoRIO's way of doing, it is the correct one for AVX, when using instr_get_src(instr, 0), we get the source0 of the instruction, and instr_get_src(instr, 1) gives the source1 (and instr_get_dst works fine aswell). On the contrary, for SSE, since the first operand is a destination AND source at the same time, DynamoRIO considers the destination as an explicit operand, and the source as implicit. That implies that the source is put at the last position in the instruction's sources, so instr_get_src(instr, 0) actually gives back the source1, while instr_get_src(instr, 1) gives source0.
From a simple logic point, this is pretty strange, but mostly, it is really problematic regarding instruction overload because one would have to manage SSE and AVX instructions differently because of DynamoRIO, not because of the Intel standard. Also, since no floating point operation is commutative, this problem actually change the values of the results which is extremely problematic.
Many instructions are impacted, here is a non limited list :
SSE -> {add|mul|sub|div}{s|p}{s|d}
AVX -> v{add|mul|sub|div}{s|p}{s|d}

Moreover, as stated in the same post, this is even more obvious for FMA instructions, because Intel manages these the same way as for SSE instructions : "instr destination(=source0), source1, source2". However, the instr for FMA is very specific for Intel, because the order of FMA is specified in the OP_code. So for instance, we can have the following :
"VFMADD231SS xmm0, xmm1, xmm2" -> multiply xmm1 by xmm2, add xmm0, put the result in xmm0.
Since the source0 is implicit, DynamoRIO places it at the end of the sources, so in third position, meaning that if don't specifically for this case, we would get the following :
"VFMADD231SS xmm0, xmm1, xmm2" becomes for DynamoRIO "VFMADD231SS xmm1, xmm2, xmm0" -> multiply xmm2 by xmm0, add xmm1, put the result in xmm0.
Obviously, this is not at all the result wanted. Now consider that there is not only one order for fma instructions, but 3 (231, 213 and 132), there are scalar and packed instructions, and instructions for double and float. So you get a whooping total of 3 * 2 * 2 = 12 special cases to take into account. Now we include FMS, NFMA, NFMA, FMAS, NFMAS, and the amount of particular cases is way too big to adapt to.

Finally, as said in the same post, this is also not consistant for PEXT instructions, as the mask is put in the second source, which is source1, and source0 (the value on which to apply the mask") is also the destination, so source0 is implicit, so DynamoRIO put it at the end of sources, meaning that if we don't adapt to this particular case, we would use source0 as the mask, not source1, and the result would be totally messed up.

Overall, the problem comes from the fact that when one of the source is also the destination, DynamoRIO considers it as implicit, and put the source/destination at the end of source. This means that one would have to change the order of the sources to have the correct result, so when we would overload instructions that don't have implicit sources, the order wouldn't be correct anymore.
I'd advice that DynamoRIO changes the way implicit operands are managed in case of these operands also being the destination, and put them at their normal position, following the Intel standard to the letter. But when one of the source is implicit but not the destination (i.e. push for instance, which only takes one explicit source, but has another implicit source, rsp), the current way of doing is fine in my opinion.
This problem is important for floating point arithmetic, as the author of the post first stated, but is even more troublesome for FMA and PEXT.

**Comment by derekbruening** [comment]
The DR IR is an abstraction and does not guarantee to match assembly conventions or vendor manuals in how operands are listed.  There are other considerations such as cross-platform generalization, hiding of unnecessary details, and facilitating common tool operations, when encoding machine code into the IR.  E.g., we considered putting the condition codes (eflags) as an explicit operand, which would not match the manuals but might improve tool code.  The point is that blindly following a particular manual's depiction is not the goal.

Additionally, each opcode has its own semantics.  The complex SIMD operations typically need special handling anyway, and the DR IR does not try to encode semantics, only sets of inputs and outputs.

That said, the point raised in this issue is valid, that it would simplify tool code to be able to treat different non-commutative opcodes the same when determining which source is used for which purpose.

To change the order, there are two considerations:

1) What was the original logic for the current order: was there a factor where it made certain tool operations nicer and changing the order would thwart that.

2) Changing the order would break existing clients, including certainly Dr. Memory and likely other clients.  They could be updated but there would be a cost.  It is probably worth paying but it is not free.

**Comment by johnfxgalea** [comment]
Another idea could be to build a middle-layered interface which provides intel operand ordering for those who want to use, while still keeping the current interface intact for backwards compatibility and future extensions (e.g. eflag operands).

**Comment by hgreving2304** [comment]
In LLVM, while most MIR probably *is* the same order as Intel's, I don't think there is a hard written down guarantee, and operands are often accessed by name. Basically similar to @johnfxgalea 's point. For noncommutative arithmetic one could have accessors like get_dividend() and get_denominator(). However, I would tend to think just making them the same order would be more straight forward.

**Comment by YValeri** [comment]
I think that changing the order to be consistent would be the best, because this problem actually isn't specified anywhere, and is actually present for all instructions using a destination as source.
I disagree with @derekbruening regarding to "blindly" following the standard : since DynamoRIO documentation isn't exactly complete, up to date or accurate, this problem is something you can only find after multiple attemps and time searching for why this thing exists. The assembly standard isn't working like that, and DynamoRIO should follow the standard to the letter, in my opinion.
Moreover, DynamoRIO is an amazing tools, and you can do a big load of dangerous stuff. So if, when doing potentially dangerous stuff, the only source on how to do it properly is the standard, but DynamoRIO doesn't follow it, you can't really be effective or efficient.

However, i do agree that regarding the current tools, building a middle layer would be better, along with a better explanation in the documentation on this "problem". Adding some functions like "instr_get_ordered_src" which would give the implicit sources at their rightful place, "instr_use_implicit_src" to tell if an instruction has this special order with implicit order, or others would be great : it would be consistant with Intel's standard and wouldn't break anything.

**Comment by hgreving2304** [comment]
@YValeri , all the developers here tend to agree and see the convenience in having the IR match Intel's assembly syntax. But - please keep in mind that DynamoRIO's IR is *not* assembly, neither is LLVM's MIR, nor is GCC's RTL etc. Making this assumption in clients is dangerous, opcodes may be combined or split, there may be other subtle differences. So while it sounds that this issue here is generally accepted with respect to particular operands, the conclusion is _not_ to exactly match assembly syntax in the IR itself.

**Comment by YValeri** [comment]
Yes i understand, and i think that actually stating this issue exists, and giving ways to work around it would be good.

**Comment by derekbruening** [comment]
> since DynamoRIO documentation isn't exactly complete, up to date or accurate

Please back up a statement like this with evidence.  This is open-source software: please submit a PR if you see an inaccuracy in the documentation.
